misc/incremental_checker.py

```diff
@@ -0,0 +1,280 @@
+            skip that commit to save time.
+        -   Cache the results after finishing.
+    4.  Rewind back to the first commit, and run mypy WITH incremental mode
+        against the `module_path` commit-by-commit, and compare to the expected
+        results found in stage 3.
+    """
+    # Stage 1: Clone repo and get ready to being testing
+    ensure_environment_is_ready(mypy_path, temp_repo_path, mypy_cache_path)
+    initialize_repo(target_repo_url, temp_repo_path, branch)
+
+    # Stage 2: Get all commits we want to test
+    if range_start.startswith("-"):
+        start_commit = get_nth_commit(temp_repo_path, int(range_start[1:]))[0]
+    else:
+        start_commit = range_start
+    commits = get_commits_starting_at(temp_repo_path, start_commit)
+
+    # Stage 3: Find and cache expected results for each commit (without incremental mode)
+    cache = load_cache(incremental_cache_path)
+    set_expected(commits, cache, temp_repo_path, module_path, mypy_cache_path)
+    save_cache(cache, incremental_cache_path)
+
+    # Stage 4: Rewind and re-run mypy (with incremental mode enabled)
+    test_incremental(commits, cache, temp_repo_path, module_path, mypy_cache_path)
+    cleanup(temp_repo_path, mypy_cache_path)
+
+
+def main() -> None:
+    help_factory = (lambda prog: RawDescriptionHelpFormatter(prog=prog, max_help_position=32))
+    parser = ArgumentParser(
+        prog='incremental_checker',
+        description=__doc__,
+        formatter_class=help_factory)
+
+    parser.add_argument("range_start", default=DEFAULT_START, metavar="COMMIT_ID_OR_NUMBER",
+                        help="the commit id to start from, or the number of "
+                        "commits to move back (see above)")
+    parser.add_argument("-r", "--repo_url", default=MYPY_REPO_URL, metavar="URL",
+                        help="the repo to clone and run tests on")
+    parser.add_argument("-m", "--module-name", default=MYPY_MODULE_NAME, metavar="NAME",
+                        help="the name of the module to typecheck")
+    parser.add_argument("--cache-path", default=CACHE_PATH, metavar="DIR",
+                        help="sets a custom location to store cache data")
+    parser.add_argument("--branch", default=None, metavar="NAME",
+                        help="check out and test a custom branch"
+                        "uses the default if not specified")
+
+    if len(sys.argv[1:]) == 0:
+        parser.print_help()
+        parser.exit()
+
+    params = parser.parse_args(sys.argv[1:])
+
+    # Make all paths absolute so we avoid having to worry about being in the right folder
+    script_path = absolute_path(sys.argv[0])
+    mypy_path = absolute_path(script_path, os.pardir, os.pardir)
+    temp_repo_path = absolute_path(mypy_path, "tmp_repo")
+    module_path = absolute_path(temp_repo_path, params.module_name)
+    incremental_cache_path = os.path.abspath(params.cache_path)
+    mypy_cache_path = absolute_path(mypy_path, "misc", ".mypy_cache")
```

Comment:
Hm, there are two other places in the code that directly reference "misc/.mypy_cache". Should probably have a single constant for that.